proxy/proxy/client_test.go

```diff
@@ -64,6 +65,31 @@ func (cs *blockingCertSource) Remote(instance string) (cert *x509.Certificate, a
 	return &x509.Certificate{}, "fake address", "fake name", "fake version", nil
 }
 
+func TestContextDialer(t *testing.T) {
+	b := &fakeCerts{}
+	c := &Client{
+		Certs: &blockingCertSource{
+			map[string]*fakeCerts{
+				instance: b,
+			},
+			forever,
+		},
+		ContextDialer: func(ctx context.Context, _, _ string) (net.Conn, error) {
+			return nil, ctx.Err()
+		},
+		Dialer: func(string, string) (net.Conn, error) {
+			return nil, fmt.Errorf("this dialer should't be used when ContextDialer is set")
+		},
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel() // already canceled
```

Comment:
Not sure I follow this logic. Can you elaborate what's going on here?

Reply:
If you're asking then it obviously needed a rework. I was passing a canceled context and checking the error to be `context.Canceled` since that's what our dialer was returning (`ctx.Err()`) although incorrectly, since `proxy.FromEnvironment()` would return that error too.

Changed it to return `errFakeDial` instead just like the `Dialer` test does.